### DISCUSSION TITLE
define:args

### DISCUSSION THREAD

**Discussion by lilnasy** [withastro/roadmap#705] [Proposal]
Exploration Branch: [withastro/compiler](https://github.com/lilnasy/compiler/compare/main...lilnasy:compiler:proposal-define-args), [withastro/astro](https://github.com/lilnasy/astro/compare/main...lilnasy:astro:proposal-define-args), _withastro/language-tools_ (pending)

# Summary
A replacement for define:vars for scripts that allows bundling.

# Background & Motivation
[`define:vars`](https://docs.astro.build/en/reference/directives-reference/#definevars) is a powerful API, but it is limited by its inability to use npm packages and other modules in the project. Unsurprisingly, users have conserved its usage to simple dark-light theme toggles.

While custom elements can be used as a [replacement in certain cases](https://docs.astro.build/en/guides/client-side-scripts/#web-components-with-custom-elements), it has historically received low adoption. Therefore, users often consider workarounds and community projects, such as [@ayco/astro-resume](https://www.npmjs.com/package/@ayco/astro-resume).

The downside frequently involves creating and keeping tracking of an identifier. This becomes problematic as the identifier must be globally unique for each instance of the component.

# `define:args`
A script using the proposed `define:args` exports a function as default. This function is called once per each usage of the its parent astro component. The lines of code surrounding the default function are executed only once per page load.
<details>
<summary>

### Defining the reusable component
</summary>

```html
<h1>Welcome to Astro</h1>
<script define:args={Astro.props.x}>
    export default function (x : string) {
        console.log(`This block reruns once for every use of the component. In this run, x is ${x}`)
    }
</script>
```
</details>
<details>
<summary>

### Usage
</summary>

```html
---
import Component from "../components/WithReusableScript.astro"
---
<Component x="Hello"/>
<Component x="from define:args"/>
```
</details>
<details>
<summary>

### Rendered HTML
</summary>

_Formatted for clarity._
```html
<!DOCTYPE html>

<h1>Welcome to Astro</h1>
<script type="module">
	import reusableFunction from "/_astro/hoisted.1e0313d1.js"
	reusableFunction(JSON.parse("\"Hello\""))
</script>

<h1>Welcome to Astro</h1>
<script type="module">
	import reusableFunction from "/_astro/hoisted.1e0313d1.js"
	reusableFunction(JSON.parse("\"from define:args\""))
</script>
```
</details>

<details>
<summary>

### Advanced Example
</summary>

```astro
---
interface Props {
    x : string
}

const x = Astro.props.x
---
<script define:args={x}>
import { variable } from "library"

console.log("Top level statements execute on the initial page load.")

export default (x : string) => {
    console.log(`This block reruns once for every use of the component. In this run, x is ${x}`)
}

document.addEventListener('astro:page-load', () => {
    console.log("This block runs on every view transition naviagtion.")
})
</script>
```
</details>

# Tentative Design
- A new type of build artifact: reusable scripts.
- Addition to the server runtime render API: `defineArgs()`.
- Compiler generates virtual module IDs for the reusable scripts. For example, `component.astro?astro&type=script&index=0&lang.ts`.
- Client build creates a mapping from virtual module IDs to real file specifiers as generated by Vite. For example, `component.astro?astro&type=script&index=0&lang.ts => /_astro/reusable.be013af6.js`
- The mapping gets included in the manifest.
- At runtime, `defineArgs()` uses the mapping to render inline scripts. For example...
```html
<script type="module">
    import reusableFunction from "/_astro/reusable.be013af6.js"
    reusableFunction(JSON.parse("\"Hello\""))
</script>
```

# Goals
- Allow module imports in a "reusable" script.

# Non Goals
- Breaking changes
- Anything directly related to styling

# Harmony
_How might this proposal interact with other in-progress proposals?_
- https://github.com/withastro/astro/pull/8004 - could inform the serialization approach used for the "args".
- https://github.com/withastro/roadmap/discussions/664 - reusable scripts are loaded only when the component is used.

**Comment by matthewp** [comment]
I love this 👏. My only note would be that it would be nice if this were somehow using the same build code as hoisted scripts, just because that's somewhat complicated and it would be good not to add another thing that's very similar.

**Reply by lilnasy** [reply]
The exploration branch does co-opt hoisted scripts' code paths at every possible point. Reusable scripts are not the same as hoisted scripts - they never go into `<head>` - but I think it would be possible to share most of the code by generalizing just a little.

**Reply by matthewp** [reply]
Ah good point about these not going in the head.

My point was more about sharing the build code, since that's split between a couple of the vite plugins so it's somewhat hard to understand. It could still be a different "concept" at the manifest level.

**Comment by matthewp** [comment]
What are the implications of this on the language tools?

**Reply by lilnasy** [reply]
Haven't looked into it yet.

**Reply by Princesseuh** [reply]
Making this not do an error is easy, making it use the proper type is hard. 

Hoisted script tags are completely separate files for the language tooling, so you'd have to somehow transpose the type between the files, which is challenging. 

Two ways I can think of:
- In the compiler, you'd need to export the types of all the elements you're passing through `define:args` so that you can import them from the virtual script tag file
- Every virtual script file could be the entire Astro component with only the script part actually mapped. That way you always fully conserve the scope. This can lead to issues if there's syntax errors in other parts of the file that affects the mapped part, but it should generally be fine (and there's way to work around it a bit)

There might be other ways, but those would be my first thought!

**Comment by matthewp** [comment]
The https://github.com/withastro/roadmap/discussions/706 proposal has an example where it checks if the function has already been called. I assume that's because the component might be rendered more than once but the script acts like a singleton. I don't see any problem with that, just want to make sure I understand that.

**Reply by lilnasy** [reply]
Yes, that's the reason.

The case where the reusable function does not need that check, is where the define:args script is authored on a top-level page (src/pages/**.astro). It would run exactly once in that case.

**Comment by lilnasy** [comment]
I think it would be good idea to use customElements instead of inline scripts. For example
```
<astro-args-37fxchfa-0 args="{ \"x\": 1, \"y\": 2 }"><astro-args-37fxchfa-0>
```
...where `37fxchfa` is the scope of the source astro file and `0` is the index of the script within it.

- Opens up to strict content security policies. The tentative design described above would have required `unsafe-inline`.
- Paves way for client directives as future work: `<script client:idle define:args={{}}>`

**Reply by matthewp** [reply]
Yeah that's interesting. In some ways the custom element approach to islands might have been a mistake as it makes the DOM more verbose and creates race conditions between the island script and directive scripts. I'm not sure we want to make that mistake again, although the CSP concern is a real one. cc @natemoo-re

**Reply by natemoo-re** [reply]
I never loved having an `astro-island` custom element, honestly. It was more of a workaround for the fact that most rendering libraries don't allow you to hydrate a subtree of the DOM so we needed to normalize hydration with a guaranteed root element.

That being said, if we need to inject an element into the DOM there isn't much of a difference between `script` and a custom element. Both break user expectations that the node tree they author will be respected in the DOM output, but Astro has already considered that an acceptable tradeoff. The CSP concern is real though, and that would be a tangible benefit to the custom element approach.

**Reply by natemoo-re** [reply]
Also, for a very long time, `astro-island` wasn't actually a defined custom element. It was just a marker. So that could be the case here as well.

**Comment by VoxelMC** [comment]
I am very interested in this idea, and I would love to help with it if it goes any further than stage 1!

**Comment by flayks** [comment]
That'd be amazing to target an Astro component node element! Like Svelte would do with `bind:this={variable}` or a `useRef()` in React.
Web components are cool but this would be more integrated and "natural". Also passing variables and other kind of props that cannot be used when using `define:vars` and `import` in the script tag for instance.

**Comment by Avataw** [comment]
I stumbled upon this very problem :)

I support this proposal - are there any workarounds you would recommend until then? :)

**Reply by delucis** [reply]
This is the approach we document: https://docs.astro.build/en/guides/client-side-scripts/#pass-frontmatter-variables-to-scripts
